src/ContentRepository/Linq/FilteredContentEnumerable.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Linq.Expressions;
+using SenseNet.ContentRepository.Storage;
+using SenseNet.Search;
+
+namespace SenseNet.ContentRepository.Linq
+{
+    public class FilteredContentEnumerable : IEnumerable<Content>
+    {
+        private readonly IEnumerable<Node> _nodeEnumerable;
+        private readonly IEnumerable<Content> _contentEnumerable;
+        private readonly SortInfo[] _sort;
+        private readonly int _top;
+        private readonly int _skip;
+        private readonly Func<Content, bool> _filter;
+
+        public int AllCount { get; private set; }
+
+        public FilteredContentEnumerable(IEnumerable nodeOrContentEnumerable, LambdaExpression filterExpression,
+            IEnumerable<SortInfo> sort, int top, int skip)
+        {
+            _nodeEnumerable = nodeOrContentEnumerable as IEnumerable<Node>;
+            _contentEnumerable = nodeOrContentEnumerable as IEnumerable<Content>;
+            if(_nodeEnumerable == null && _contentEnumerable == null)
+                throw new InvalidOperationException(
+                    "The 'nodeOrContentEnumerable' can be IEnumerable<Node> or IEnumerable<Content>.");
+
+            _sort = sort == null ? new SortInfo[0] : sort.ToArray();
+            _top = top;
+            _skip = skip;
+
+            if(filterExpression != null)
+            {
+                var func = filterExpression.Compile();
+                _filter = func as Func<Content, bool>;
+                if (_filter == null)
+                    throw new InvalidOperationException(
+                        "Invalid filterExpression (LambdaExpression): return value must be bool, " +
+                        "parameter need to be " + typeof(Node).FullName);
```

Comment:
The error message is inconsistent with the expected parameter type of `Content`.
Also please change the message to 'needs to be'.